cmake: enable LUAJIT_ENABLE_CHECKHOOK option

igormunkin opened the pull request:
In scope of this commit check for instruction/line hooks for compiled code is turned on. As a result a special pad that validates whether debug hook is set while trace is being executed.

Closes #7762

NO_DOC=internal

Signed-off-by: Igor Munkin <imun@tarantool.org>

Buristan commented on test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua:
```diff
@@ -0,0 +1,69 @@
+    -- the scope of the debug hook below.
+    local lidx, lval
+    -- Table with lines to be hit by debug hook below.
+    local lines = {
```
Side note: I really hope that we will have no need to patch this tests (like we do for memprof).

igormunkin replied:
@fckxorg, nice, I'll rewrite the chunk this way.

igormunkin replied:
I have come up with the following change. If you have any better ideas, feel free to share them.

```diff
diff --git a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
index af48479df..242e3f8a1 100644
--- a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
+++ b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
@@ -4,13 +4,31 @@ local g = t.group()
 
 -- Test payload to be recorded. It calculates the arithmetic sum
 -- of elements from 1 to the given n.
-local function sumN(n)
-    local result = 0
-    for i = 1, n do
-        result = result + i
+-- XXX: To avoid another fragile test, use loadstring to generate
+-- the function with code motion agnostic line numbers to be used
+-- in hook validation below.
+local sumN = assert(load([[
+    return function(n)
+        local result = 0
+        for i = 1, n do
+            result = result + i
+        end
+        return result
     end
-    return result
-end
+]]))()
+
+-- Table with lines to be hit by debug hook below.
+local lines = {
+    2,  -- local result = 0
+    3,  -- for i = 1, n do
+    4,  --   result = result + i -- (i is 1)
+    3,  -- for i = 1, n do
+    4,  --   result = result + i -- (i is 2)
+    3,  -- for i = 1, n do
+    4,  --   result = result + i -- (i is 3)
+    3,  -- for i = 1, n do
+    6,  -- return result
+}
 
 g.test_luajit_enable_checkhook = function()
     -- Setup JIT engine to record a trace as soon as possible.
@@ -30,30 +48,34 @@ g.test_luajit_enable_checkhook = function()
     -- Return JIT engine setup to its defaults.
     jit.opt.start('hotloop=56')
 
-    -- Upvalues to fetch key and value from lines to be used in
-    -- the scope of the debug hook below.
-    local lidx, lval
-    -- Table with lines to be hit by debug hook below.
-    local lines = {
-        60, -- local sum = sumN(3)
-        8,  -- local result = 0
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 1)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 2)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 3)
-        9,  -- for i = 1, n do
-        12, -- return result
-        63, -- debug.sethook()
-    }
+    -- Upvalue for a key from lines to be used in the scope of the
+    -- the debug hook below.
+    local lidx = 1
+
-    -- Upvalues to fetch key and value from lines to be used in
-    -- the scope of the debug hook below.
-    local lidx, lval
-    -- Table with lines to be hit by debug hook below.
-    local lines = {
-        60, -- local sum = sumN(3)
-        8,  -- local result = 0
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 1)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 2)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 3)
-        9,  -- for i = 1, n do
-        12, -- return result
-        63, -- debug.sethook()
-    }
+    -- Upvalue for a key from lines to be used in the scope of the
+    -- the debug hook below.
+    local lidx = 1
+
+    -- XXX: The only fragile part of this test is here. After the
+    -- hook is set few lines below, it is fired two more times in
+    -- addition to all entries in <lines> table above. Hence we
+    -- have to extend <lines> table with the following entries.
+    -- 1. The call of the compiled test payload below.
+    -- local sum = sumN(3)
+    -- In this case the hook is fired before any of the lines
+    -- within <sumN> function, so the entry is added at the least
+    -- possible index in the <lines> (i.e. <lidx> starting value).
+    table.insert(lines, lidx, 82)
+    -- 2. Turning of the hook after the test payload returns.
+    -- debug.sethook()
+    -- This is the last time the hook is fired, so the entry is
+    -- added to the end of the <lines> table.
+    table.insert(lines, 85)
 
     -- Set debug hook.
     debug.sethook(function(hookname, line)
```

igormunkin replied:
Finally I came up with the following: there is no need to check the first and the last entry of `lines` table, hence, just ignore the assertion when the hook is fired for these cases. The diff is below:
```diff
diff --git a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
index e3dea7c60..ee76bf7ee 100644
--- a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
+++ b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
@@ -50,30 +50,32 @@ g.test_luajit_enable_checkhook = function()
 
     -- Upvalue for a key from lines to be used in the scope of the
     -- the debug hook below.
-    local lidx = 1
+    local lidx = 0
 
-    -- XXX: The only fragile part of this test is here. After the
-    -- hook is set few lines below, it is fired two more times in
-    -- addition to all entries in <lines> table above. Hence we
-    -- have to extend <lines> table with the following entries.
-    -- 1. The call of the compiled test payload below.
-    -- local sum = sumN(3)
-    -- In this case the hook is fired before any of the lines
-    -- within <sumN> function, so the entry is added at the least
-    -- possible index in the <lines> (i.e. <lidx> starting value).
-    table.insert(lines, lidx, 82)
-    -- 2. Turning of the hook after the test payload returns.
-    -- debug.sethook()
-    -- This is the last time the hook is fired, so the entry is
-    -- added to the end of the <lines> table.
-    table.insert(lines, 85)
-
-    -- Set debug hook.
+    -- XXX: After the hook is set, it is fired two more times in
+    -- addition to all entries in <lines> table above:
+    -- 1. The call of the compiled test payload below:
+    --         local sum = sumN(3)
+    --   In this case the hook is fired before any of the lines
+    --   within <sumN> function, so we can ignore it by setting
+    --   <lidx> initial value to zero. As a result the lookup
+    --   in <lines> table yields nil for this case.
+    -- 2. Turning of the hook after the test payload returns:
+    --         debug.sethook()
+    --   This is the last time the hook is fired, and <lidx> value
+    --   equals to #lines + 1. As a result the lookup also yields
+    --   nil for this case.
+    -- Since the test is aimed to check whether the hook is fired
+    -- on trace compiled for the payload above, there is no need
+    -- to check the actual value of the line for the both
+    -- aforementioned cases and we can simply skip the assertion.
     debug.sethook(function(hookname, line)
         -- XXX: Only line hooks are fired.
         assert(hookname == 'line', 'Unexpected hook type is fired')
         local lval = lines[lidx]
-        t.assert_equals(line, lval)
+        if lval then
+            t.assert_equals(line, lval)
+        end
         lidx = lidx + 1
     end, 'l')
```

Buristan commented on test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua:
```diff
@@ -0,0 +1,69 @@
+    debug.sethook(function(hookname, line)
+        -- XXX: Only line hooks are fired.
+        assert(hookname == 'line', 'Unexpected hook type is fired')
+        lidx, lval = next(lines, lidx)
```
I'm really against such `next` usage here.

From the [Lua 5.1 Reference Manual](https://www.lua.org/manual/5.1/manual.html#pdf-next):
> The order in which the indices are enumerated is not specified, even for numeric indices. (To traverse a table in numeric order, use a numerical `for` or the [`ipairs`](https://www.lua.org/manual/5.1/manual.html#pdf-ipairs) function.) 

So, this test is fragile regarding internal `next` implementation. 
I suggest to use just numeric index incremented on each hook call to fetch line value from the table.

igormunkin replied:
Be-be-be, fine I'll rewrite it via numeric upvalue.

Buristan commented on test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua:
```diff
@@ -0,0 +1,69 @@
+    jit.flush()
+
+    -- Check there is no traces at the beginning of the test.
+    assert(not checktrace(1), "JIT flush doesn't work")
```
Minor: we may mention, that `checktrace()` is a LuaC function, so no traces will be compiled for it here.

igormunkin replied:
I believe this is quite clear for ones who is concerned about this fact, and I doubt this particular comment you're suggesting above make something clear to the others.

Ignoring.

igormunkin commented:
Sergey,

Thanks for your review!

> Hi, Igor! Thanks for the patch! Please consider my comments below.
> 
> > cmake: enable LUAJIT_ENABLE_CHECKHOOK option
> 
> Maybe it should be `build:` or `luajit:` tag here? Just suggestions, feel free to ignore.

I studied `git log` entries and decided to leave it as is (consider be5c104 for example).
Ignoring.

> 
> > In scope of this commit check for instruction/line hooks for compiled
> 
> Typo: s/in scope/in the scope/

Fixed, thanks!

> 
> > code is turned on. As a result a special pad is emitted to the trace to
> > validate whether debug hook is set while this trace is being executed.
> 
> Typo: s/debug hook/a debug hook/

Fixed also.

> 
> Also, I suggest to mention that LuaJIT submodule is bumped with corresponding commits.

LuaJIT submodule will be bumped out of the scope of this PR. The existing bump is just for tests.

igormunkin commented:
Sergey,

Thanks for your review!

> Hi, Igor! Thanks for the fixes! LGTM, except a single nit regarding the test.

Added your tag:
| Reviewed-by: Sergey Kaplun <skaplun@tarantool.org>

BTW, I also need to rebase the patch on master to remove submodule bump from the PR.

igormunkin commented on test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua:
```diff
@@ -0,0 +1,91 @@
+        -- XXX: Only line hooks are fired.
+        assert(hookname == 'line', 'Unexpected hook type is fired')
+        local lval = lines[lidx]
+        print(lidx, lval, line)
```
Here is the diff:
```diff
diff --git a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
index 242e3f8a1..e3dea7c60 100644
--- a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
+++ b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
@@ -73,7 +73,6 @@ g.test_luajit_enable_checkhook = function()
         -- XXX: Only line hooks are fired.
         assert(hookname == 'line', 'Unexpected hook type is fired')
         local lval = lines[lidx]
-        print(lidx, lval, line)
         t.assert_equals(line, lval)
         lidx = lidx + 1
     end, 'l')
     ```

igormunkin commented:
Max,

Thanks for your review!

> LGTM now

Added your tag:
| Reviewed-by: Maxim Kokryashkin <m.kokryashkin@tarantool.org>
